include/arch/arm/armv/armv8-a/64/armv/vcpu.h

```diff
@@ -624,11 +624,23 @@ static inline void armv_vcpu_boot_init(void)
     /* set the SCTLR_EL1 for running native seL4 threads */
     MSR(REG_SCTLR_EL1, SCTLR_EL1_NATIVE);
     isb();
+
+#if defined(ARM_HYP_TRAP_CP14_IN_VCPU_THREADS) || defined(ARM_HYP_TRAP_CP14_IN_NATIVE_USER_THREADS)
+    initHDCR();
+#endif
 }
 
 static inline void armv_vcpu_save(vcpu_t *vcpu, UNUSED bool_t active)
 {
     vcpu_save_reg_range(vcpu, seL4_VCPUReg_TTBR0, seL4_VCPUReg_SPSR_EL1);
+
+#ifdef ARM_HYP_CP14_SAVE_AND_RESTORE_VCPU_THREADS
+    /* This is done when we are asked to save and restore the CP14 debug context
+     * of VCPU threads; the register context is saved into the underlying TCB.
+     */
+    saveAllBreakpointState(vcpu->vcpuTCB);
+#endif
+    isb();
```

Comment:
This `isb()` is visible outside the `#ifdef` -- was that intentional?

Reply:
It might not be necessary on ARMv8, as the original code didn't have it. This was probably a mistake from when I was adapting the equivalent ARMv7 code, which does have an isb() outside of the #ifdef in this function.

Reply:
Although I'm not sure why it is needed on armv7. Do you have any idea?

Reply:
No, it doesn't really make sense to me. It does make sense (maybe) after breakpoint stuff, but not after just saving registers.

Reply:
Yeah, I can kind of see why you might want to do something like this after enabling/disabling breakpoints, but that's not what saveAllBreakpointState does anyway. I will remove the isb() from the ARMv8 version but leave the ARMv7 version as it is for now.

Reply:
I've been looking into the git history a bit, and the `isb()` has been there after or inside `vcpu_save` for armv7 since at least the open source release. There is a `dsb()` somewhere at the beginning of the `vcpu_save` chain, which makes sense to me, but the `isb()` still does not (especially after doing the save).